src/ingest/SpecRepo.cpp

```diff
@@ -151,15 +173,23 @@ void genSpecs4Roll(const std::string& version,
     // A roll spec will cover X days given table location of source data
     const auto now = Evidence::now();
     const auto max_days = table->max_hr / 24;
-    for (size_t i = 0; i <= max_days; ++i) {
-      // we only provide single macro for now
+    size_t i;
+    for (i = 0; i <= max_days; ++i) {
+      // run hourly partition list
+      for(size_t j = 0 ; j < DAY_HOURS; j++) {
+        genSpecs4HourlyRoll(version, table, 24, i, specs);
+      }
+      // TODO: refactor to have better support export from HMS chenqin
+      // run daily partition list
       auto timeValue = now - i * DAY_SECONDS;
```

Comment:
I feel we may need to introduce a way to tell how many different macros are needed to gen file specs.

s3://zzz/%date%/ => [date]
s3://zzz/ds=%date%/hr=%hour%/ => [date, hour]

And we can maintain a list of macro definition internally, this needs to be refactored from single date value into a property bag or an object definition to support all macros:  https://github.com/varchar-io/nebula/blob/2303e4f11e40eca1566ed20c83fd5223a2efdc9d/src/ingest/IngestSpec.h#L68

Reply:
And based on different macro set - we may have different loops.